doc/src/manual/interfaces.md

```diff
@@ -371,3 +371,189 @@ If you are defining an array type that allows non-traditional indexing (indices
+
+## Specializing broadcasting
+
+| Methods to implement | Brief description |
+|:-------------------- |:----------------- |
+| `Broadcast.rule(::Type{SrcType}) = ContainerType` | Output type produced by broadcasting `SrcType` |
+| `similar(f, r::Broadcast.Result{ContainerType}, As...)` | Allocation of output container |
+| **Optional methods** | | |
+| `Broadcast.rule(::Type{ContainerType1}, ::Type{ContainerType2}) = ContainerType` | Precedence rules for output type |
+| `Broadcast.indices(::Type, A)` | Declaration of the indices of `A` for broadcasting purposes (for AbstractArrays, defaults to `Base.indices(A)`) |
+| **Bypassing default machinery** | |
+| `broadcast(f, As...)` | Complete bypass of broadcasting machinery |
+| `broadcast(f, r::Broadcast.Result{ContainerType,Void,Void}, As...)` | Bypass after container type is computed |
+| `broadcast(f, r::Broadcast.Result{ContainerType,ElType,<:Tuple}, As...)` | Bypass after container type, eltype, and indices are computed |
+
+[Broadcasting](@ref) is triggered by an explicit call to `broadcast` or `broadcast!`, or implicitly by
+"dot" operations like `A .+ b`. Any `AbstractArray` type supports broadcasting,
+but the default result (output) type is `Array`. To specialize the result for specific input type(s),
+the main task is the allocation of an appropriate result object.
+(This is not an issue for `broadcast!`, where
+the result object is passed as an argument.) This process is split into two stages: computation
+of the type from the arguments ([`Broadcast.rule`](@ref)), and allocation of the object
+given the resulting type with a broadcast-specific ([`similar`](@ref)).
+
+`Broadcast.rule` is somewhat analogous to [`promote_rule`](@ref), except that you
+may only need to define a unary variant. The unary variant simply states that you intend to
+handle broadcasting for this type, and do not wish to rely on the default fallback. Most
+implementations will be simple:
+
+```julia
+Broadcast.rule(::Type{<:MyType}) = MyType
+```
+where unary `rule` should typically discard type parameters so that any binary `rule` methods
+can be concrete (without using `<:` for type arguments).
+
+For `AbstractArray` types, this prevents the fallback choice, `Broadcast.BottomArray`,
+which is an `AbstractArray` type that "loses" to every other `AbstractArray` type in a binary call
+`Broadcast.rule(S, T)` for two types `S` and `T`.
+You do not need to write a binary `rule` unless you want to establish precedence for
+two or more non-`BottomArray` types. If you do write a binary rule, you do not need to
+supply the types in both orders, as internal machinery will try both. For more detail,
+see [below](@ref writing-binary-broadcasting-rules).
+
+The actual allocation of the result array is handled by specialized implementations of `similar`:
+
+```julia
+Base.similar(f, r::Broadcast.Result{ContainerType}, As...)
+```
+
+`f` is the operation being performed, `ContainerType` signals the resulting container type
```

Comment:
Should there be an "and" following the comma, or should the period terminating this sentence instead be a comma, or am I misreading these sentences? :)